routes.py

```diff
@@ -0,0 +1,193 @@
+
+
+class AccessTokenResponse(messages.Message):
+    token = messages.StringField(1)
+    expires_at = message_types.DateTimeField(2)
+
+
+class ServiceAccountKeyResponse(messages.Message):
+    data = JsonField(1)  # not quite right as it drops the json under a data element
+
+
+class ServiceAccountAccessTokenResponse(messages.Message):
+    token = messages.StringField(1)
+
+
+class SubSystemStatusResponse(messages.Message):
+    ok = messages.BooleanField(1)
+    message = messages.StringField(2)
+    subsystem = messages.StringField(3)
+
+
+class StatusResponse(messages.Message):
+    ok = messages.BooleanField(1)
+    subsystems = messages.MessageField(SubSystemStatusResponse, 2, repeated=True)
+
+
+class AuthorizationUrlResponse(messages.Message):
+    url = messages.StringField(1)
+
+
+config = ConfigParser.ConfigParser()
+config.read("config.ini")
+
+
+def create_provider(provider_name):
+    client_id = config.get(provider_name, 'CLIENT_ID')
+    client_secret = config.get(provider_name, 'CLIENT_SECRET')
+    open_id_config_url = config.get(provider_name, 'OPEN_ID_CONFIG_URL')
+    fence_base_url = config.get(provider_name, 'FENCE_BASE_URL')
+    user_name_path_expr = config.get(provider_name, 'USER_NAME_PATH_EXPR')
+
+    sam_base_url = config.get('sam', 'BASE_URL')
+
+    extra_authz_url_params = {}
+    extra_params_key = 'EXTRA_AUTHZ_URL_PARAMS'
+    if config.has_option(provider_name, extra_params_key):
+        extra_params_raw = config.get(provider_name, extra_params_key)
+        extra_authz_url_params = ast.literal_eval(extra_params_raw)
+
+    open_id_config = OpenIdConfig(provider_name, open_id_config_url)
+    oauth_adapter = OauthAdapter(client_id, client_secret, open_id_config, provider_name)
+    fence_api = FenceApi(fence_base_url)
+    sam_api = SamApi(sam_base_url)
+
+    fence_tvm = FenceTokenVendingMachine(fence_api, sam_api, oauth_adapter, provider_name)
+    return BondProvider(fence_tvm, Bond(oauth_adapter,
+                                        fence_api,
+                                        sam_api,
+                                        fence_tvm,
+                                        provider_name,
+                                        user_name_path_expr,
+
+                                        extra_authz_url_params))
+
+
+def _get_provider(self, provider_name):
+    if provider_name in self.providers:
+        return self.providers[provider_name]
+    else:
+        raise exceptions.NotFound("provider {} not found".format(provider_name))
+
+
+class BondProvider:
+    def __init__(self, fence_tvm, bond):
+        self.fence_tvm = fence_tvm
+        self.bond = bond
+
+
+routes = Blueprint('bond', __name__, '/')
+
+bond_providers = {provider_name: create_provider(provider_name)
+                  for provider_name in config.sections() if provider_name != 'sam'}
+auth = authentication.Authentication(authentication.default_config())
+
+
+@routes.route('/api/link/v1/providers', methods=["GET"])
```

Comment:
You asked me about this already, but it'd be nice to have some kind of route hierarchy rather than having a flat set of routes. It doesn't look like Flask really supports that, but it might be nice to set `'/api/link/v1'` equal to `api_routes_base` or something like that so that each of these routes would be a bit more grouped at least.  
```suggestion
api_routes_base = '/api/link/v1'
@routes.route(api_routes_base + '/providers', methods=["GET"])
```